xarray/core/combine.py

```diff
@@ -420,6 +66,82 @@ def _infer_tile_ids_from_nested_list(entry, current_pos):
         yield current_pos, entry
 
 
+def _infer_concat_order_from_coords(datasets):
+
+    concat_dims = []
+    tile_ids = [() for ds in datasets]
+
+    # All datasets have same variables because they've been grouped as such
+    ds0 = datasets[0]
+    for dim in ds0.dims:
+
+        # Check if dim is a coordinate dimension
+        if dim in ds0:
+
+            # Need to read coordinate values to do ordering
+            coord_vals = [ds[dim].values for ds in datasets]
+
+            # If dimension coordinate values are same on every dataset then
+            # should be leaving this dimension alone (it's just a "bystander")
+            if not _all_arrays_equal(coord_vals):
```

Comment:
I would suggest doing these checks with pandas.Index objects instead, e.g.,
```python
indexes = [ds.indexes.get(dim) for ds in datasets]
if any(index is None for index in indexes):
    raise ValueError('every dimension needs a coordinate for inferring concatation order')

# this assumes we have at least one dataset, but I think that is safe?
if not all(index.equals(indexes[0]) for index in indexes[1:]):
    concat_dims.append(dim)

    if not all(index.is_monotonic_increasing for index in indexes):
        raise ValueError
    
    first_items = indexes[0][:1].concat([index[:1] for index in indexes])

    if first_items.size < len(indexes):
        raise ValueError("can't handle size zero dimensions")

    order = first_items.to_series().rank(method='dense').astype(int).values - 1
```

Reply:
pandas `Index` objects don't seem to actually have a `concat()` method but apart from that I've put this in now, thanks.

This approach seems to also work for string coords and datetime coords, but I'm not sure I'm that's guaranteed pandas behaviour (we are using `.rank()` on non-numeric quantities), and it also is sorting the strings lexicographically, not naturally, which arguably isn't what we want (e.g. for a coordinate dimension like `['run8', 'run9', 'run10']`).

Reply:
Oops, it's `pandas.Index.append`, not `concat`. The virtue of using `append()` is that it should preserve the exact index type.

I think lexciographic order is probably the right choice for us. Natural sorting sort of makes sense for print output for users, but it's pretty convoluted / hard to debug so I doubt it makes sense to internally in xarray.

Reply:
> Oops, it's pandas.Index.append, not concat. The virtue of using append() is that it should preserve the exact index type.

I did it with by initialising a new `pandas.Index` object instead, which personally I think is a bit clearer. I guess that might not preserve the type though?

> I think lexicographic order is probably the right choice for us. Natural sorting sort of makes sense for print output for users, but it's pretty convoluted / hard to debug so I doubt it makes sense to internally in xarray.

:+1: